### PR TITLE
Allow wasm-xlsxwriter to be used from both from web and nodejs

### PR DESCRIPTION
**Pull request by henrikekblad**

### CODE REVIEW

**Comment by kenkoooo** [comment]
Thank you for your contribution! Could you provide guidance on how we can test it to ensure it works as expected?

File: README.md [hunk 1]
```diff
@@ -82,6 +82,72 @@ worksheet.insertImage(1, 2, image);
...
+	// Freeze header
+	worksheet.setFreezePanes(1, 0);
+
+	// Add autofilter to header
+	worksheet.autofilter(0, 0, rows.length, header.length - 1);
+
+	// Return buffer with xlsx data
+	return workbook.saveToBufferSync();
+});
+```
```
Comment (tamaroning) [inline README.md:148]: I found some type errors in this example. I'd love to see type script code here too.

```suggestion
```ts
import { Color, DocProperties, Format, FormatAlign, Workbook } from "wasm_xlsxwriter";

/**
 * Demo function that create a xlsx buffer from a header array and data rows 
 *
 * @param {string[]} header - Sheet header strings
 * @param {(string | number)[][]} rows - Array of arrays containing sheet rows
 * @returns {Buffer} Excel file data
 */
function writeExcel(header: string[], rows: (string | number)[][]): Buffer {
    // Create a new Excel file object.
    const workbook = new Workbook();

    // Set a doc property
    const props = new DocProperties()
        .setCompany('Test, Inc.');
    workbook.setProperties(props);

    // Add a worksheet with name to the workbook.
    const worksheet = workbook.addWorksheet();
    worksheet.setName('Export');

    // Create some formats to use in the worksheet.
    const headerStyle = new Format();
    headerStyle
        .setAlign(FormatAlign.Top)
        .setTextWrap()
        .setBackgroundColor(Color.red());

    // Write sheet header
    worksheet.writeRowWithFormat(0, 0, header, headerStyle)

    // Write sheet data
    worksheet.writeRowMatrix(1, 0, rows);

    // Autofit columns
    worksheet.autofit();

    // Freeze header
    worksheet.setFreezePanes(0, 0);

    // Add autofilter to header
    worksheet.autofilter(0, 0, rows.length, header.length - 1);

    // Return buffer with xlsx data
    const uint8Array = workbook.saveToBufferSync();
    return Buffer.from(uint8Array);
}
```
```

File: package.json [hunk 2]
```diff
@@ -2,20 +2,24 @@
...
+        "rust/pkg/nodejs/wasm_xlsxwriter.d.ts"
     ],
     "license": "MIT",
     "repository": {
         "type": "git",
         "url": "https://github.com/estie-inc/wasm-xlsxwriter"
     },
     "scripts": {
-        "build": "RUSTFLAGS='-C target-feature=+bulk-memory' wasm-pack build ./rust --target web --release && rm rust/pkg/.gitignore",
+        "build": "RUSTFLAGS='-C target-feature=+bulk-memory' wasm-pack build ./rust --target nodejs --out-dir pkg/nodejs --release && wasm-pack build ./rust --target web --out-dir pkg/web --release && rm rust/pkg/web/.gitignore && rm rust/pkg/nodejs/.gitignore",
```
Comment (tamaroning) [inline package.json:22]: This should be separate script commands

Comment (henrikekblad) [reply package.json:22]: I guess only you have access to the github workflows? Maybe it's better you do the separation after the PR has been merged?

**Review by kenkoooo — COMMENTED** [review]
It seems that we need to put more effort into supporting both browsers and Node runtimes. For example, [brotli-wasm](https://github.com/httptoolkit/brotli-wasm/) has its own build script to achieve that. The [bundler](https://rustwasm.github.io/wasm-bindgen/reference/deployment.html?highlight=bundler#bundlers) target can be an option, but according to the documentation, only Webpack supports this feature at this time.

**Comment by henrikekblad** [comment]
Ok, I'll follow the brotli example. They seem to be working on an `inline` target that may be an option in the future.

For reference:
https://github.com/rustwasm/wasm-pack/issues/1334
https://github.com/rustwasm/wasm-bindgen/pull/4065

**Comment by kenkoooo** [comment]
It seems that your change no longer supports the web, such as Next.js. I added a test in #26 to ensure that it still works with Next.js. I hope it might be helpful.

**Comment by henrikekblad** [comment]
Ok thanks, cross-env seems to fix the problem @kenkoooo.

**Comment by kenkoooo** [comment]
I'm really sorry, but the CI setup was incorrect. It's already been fixed in #28. Would you mind retrying with the latest version? I apologize for the inconvenience, and thank you for your understanding.

**Comment by henrikekblad** [comment]
Are you sure about the rollback @kenkoooo? I actually experienced problem with #26 when building locally without the cross-env fix.

**Comment by kenkoooo** [comment]
The latest main branch passes all tests. Could you please rebase your branch onto it? Thank you.

**Comment by henrikekblad** [comment]
Could this be related?

https://github.com/vercel/next.js/discussions/61921

**Comment by kenkoooo** [comment]
I'm not entirely sure how Next.js bundles libraries, but some libraries, such as [nextjs-auth0](https://github.com/auth0/nextjs-auth0), provide both client-side and server-side functionalities.

File: examples/nextjs/app/page.tsx [hunk 3]
```diff
@@ -1,5 +1,5 @@
 "use client";
-import initModule, { Workbook } from "wasm-xlsxwriter";
+import initModule, { Workbook } from "../../../pkg.web/";
```
Comment (kenkoooo) [inline examples/nextjs/app/page.tsx:2]: Would you mind avoiding the use of absolute paths and using package names instead? We'd like to ensure that the package can be imported correctly. Could you please use something like `wasm-xlsxwriter/client`?

**Review by kenkoooo — CHANGES_REQUESTED** [review]

**Comment by henrikekblad** [comment]
Ok, you guys seemed to run a newver version of npm. When I updated npm and re-created package-lock-file it seems to work.

**Comment by kenkoooo** [comment]
Thank you for your excellent work! I’ll review and merge it this week.

**Pull request merged**